library_generation/model/config_change.py

```diff
@@ -0,0 +1,159 @@
+from library_generation.utils.proto_path_utils import get_proto_paths
+
+
+class ChangeType(Enum):
+    GOOGLEAPIS_COMMIT = 1
+    REPO_LEVEL_CHANGE = 2
+    LIBRARIES_ADDITION = 3
+    # As of Mar. 2024, we decide not to produce this type of change because we
+    # still need to manually remove the libray.
+    # LIBRARIES_REMOVAL = 4
+    LIBRARY_LEVEL_CHANGE = 5
+    GAPIC_ADDITION = 6
+    # As of Mar. 2024, we decide not to produce this type of change because we
+    # still need to manually remove the libray.
+    # GAPIC_REMOVAL = 7
+
+
+class HashLibrary:
+    """
+    Data class to group a LibraryConfig object and its hash value together.
+    """
+
+    def __init__(self, hash_value: int, library: LibraryConfig):
+        self.hash_value = hash_value
+        self.library = library
+
+
+class LibraryChange:
+    def __init__(self, changed_param: str, latest_value: str, library_name: str = ""):
+        self.changed_param = changed_param
+        self.latest_value = latest_value
+        self.library_name = library_name
+
+
+class QualifiedCommit:
+    def __init__(self, commit: Commit, libraries: set[str]):
+        self.commit = commit
+        self.libraries = libraries
+
+
+class ConfigChange:
+    ALL_LIBRARIES_CHANGED = None
+
+    def __init__(
+        self,
+        change_to_libraries: dict[ChangeType, list[LibraryChange]],
+        baseline_config: GenerationConfig,
+        latest_config: GenerationConfig,
+    ):
+        self.change_to_libraries = change_to_libraries
+        self.baseline_config = baseline_config
+        self.latest_config = latest_config
+
+    def get_changed_libraries(self) -> Optional[list[str]]:
+        """
+        Returns a unique, sorted list of library name of changed libraries.
+        None if there is a repository level change, which means all libraries
+        in the latest_config will be generated.
+        :return: library names of change libraries.
+        """
+        if ChangeType.REPO_LEVEL_CHANGE in self.change_to_libraries:
+            return ConfigChange.ALL_LIBRARIES_CHANGED
+        library_names = set()
+        for change_type, library_changes in self.change_to_libraries.items():
+            if change_type == ChangeType.GOOGLEAPIS_COMMIT:
+                library_names.update(self.__get_library_names_from_qualified_commits())
+            else:
+                library_names.update(
+                    [library_change.library_name for library_change in library_changes]
+                )
+        return sorted(list(library_names))
+
+    def get_qualified_commits(
+        self,
+        repo_url: str = "https://github.com/googleapis/googleapis.git",
+    ) -> list[QualifiedCommit]:
+        """
+        Returns qualified commits from configuration change.
+
+        A qualified commit is a commit that changes at least one file (excluding
+        BUILD.bazel) within a versioned proto path in the given proto_paths.
```

Comment:
I know we made the decision to exclude BUILD.bazel previously, but we may need to revisit it for scenarios that the changes include transport/numeric_enums/include_samples. It means we may have to parse BUILD.bazel in this step, which would definitely make it more complicated. It was fine because we regenerate the whole repo anyway, but if we start excluding libraries from regeneration, then we need to be more careful.

Reply:
I think we exclude `BUILD.bazel` because we don't want changes in other languages appear in commit messages, e.g., `[ruby] xxx`.

How about I change the regex in filtering commit message to exclude irrelevant message, rather than excluding from generation.

Reply:
> I think we may have to parse the BUILD.bazel file to see if there is any changes to transport/numeric_enums/include_samples in Java targets. Of course this can be a separate a PR and slightly lower priority as they don't change often.

I agree. I'll add a task as the next step and keep this part as-is.